### PR TITLE
Exported `TradeOffer` class in `steam-tradeoffer-manager`

### PR DESCRIPTION
**Pull request by Baterka**
This class is not exported for some weird reason. It prevents further extending of the class (e.g. creating promise wrapper in my case)

Additional info:

Both js modules are exported, but just the `TradeOfferManager` have exported type declaration:
https://github.com/DoctorMcKay/node-steam-tradeoffer-manager/blob/master/lib/index.js
https://github.com/DoctorMcKay/node-steam-tradeoffer-manager/blob/master/lib/classes/TradeOffer.js

### CODE REVIEW

**Review by kldzj — CHANGES_REQUESTED** [review]
Hey, thanks for your PR. 

This approach will not work as `TradeOffer` is not exported in `lib/index.js`. But it should be possible to [add an export for the `lib/classes/TradeOffer.js`](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html#library-file-layout).

**Comment by kldzj** [comment]
It'd also be great if you could add some basic usage of the typed `TradeOffer` class using the following import to the test file. :)

```ts
import TradeOffer = require('steam-tradeoffer-manager/lib/classes/TradeOffer'); // append .js extension if needed
```

**Comment by Baterka** [comment]
> It'd also be great if you could add some basic usage of the typed `TradeOffer` class using the following import to the test file. :)
> 
> ```ts
> import TradeOffer = require('steam-tradeoffer-manager/lib/classes/TradeOffer'); // append .js extension if needed
> ```

I was testing it now. I created the `classes/TradeOffer.d.ts` file and I was doing this in my project:
`import TradeOffer from 'steam-tradeoffer-manager/classes/TradeOffer.js'`
which is very similar to already working
`import CEconItem from 'steamcommunity/classes/CEconItem.js'` (similar library from same author)
but I am getting `TS1192: Module '"node_modules/@types/steam-tradeoffer-manager/classes/TradeOffer"' has no default export.`
So II think we can just close the PR as my knowledge is probably not good enough to do it properly :D

**Comment by kldzj** [comment]
You're very much welcome to keep trying. That's how we learn and grow, right? :) If you commit your changes I can try to look for issues.

Otherwise feel just as free to close this PR and I might see what I can do in the following days.

**Comment by kldzj** [comment]
Please also note that `steamcommunity` has a different project structure which does not include the `lib/` dir.

Perhaps I'm wrong, but the import paths in your error look wrong.

**Comment by kjsmita6** [comment]
Related to this, I also have a massive change set in progress that I
attempted to include this. Might separate that from the rest of my changes
as I haven't figured out a good way to do it that I'm satisfied with.

**Comment by kldzj** [comment]
> Related to this, I also have a massive change set in progress that I attempted to include this. Might separate that from the rest of my changes as I haven't figured out a good way to do it that I'm satisfied with.

Good to hear!

In theory it should be relatively straightforward. For the `lib/classes/` files we would have to mimic the path for the new `.d.ts` files from the original package and import them in the `index.d.ts` respectively.

Then we should be able to import the typed classes using the relative path.

**Comment by Baterka** [comment]
@kldzj This will work :) Tested it in project.

**Comment by Baterka** [comment]
Frick...

**Comment by kldzj** [comment]
Don't worry, currently checking as well what's wrong. It seems like some reference just broke, but I'll have to verify to be sure. The approach now looks pretty much correct. :)

**Comment by Baterka** [comment]
Yea. I created common.d.ts. Will try local build and then I will commit.

**Comment by kldzj** [comment]
Alright, so, only small changes should be required to make this work. :)

1. Move the `InventoryCallback` type to the `TradeOfferManager` namespace
image

2. Change references in `index.d.ts`
image

3. Change references in `lib/classes/TradeOffer.d.ts`
image


To verify that all is working correctly you can run the tests yourself:

1. Install dependencies in the DefinitelyTyped repo (`npm install` in the project root)
2. I had an issue running the `npm test steam-tradeoffer-manager` command to run
3. `npm install --prefix /tmp/dt/types/bytebuffer`
4. Now the tests should work (again `npm test steam-tradeoffer-manager`)

**Comment by kldzj** [comment]
> Yea. I created common.d.ts. Will try local build and then I will commit.

I thought about making a shared types file as well, but it seems more reasonable to just put it in the already used namespace I think. :)

**Comment by kldzj** [comment]
While you're at it could you please add an example usage to our `steam-tradeoffer-manager-tests.ts`? Perhaps just some basic class extending from `TradeOffer`.

**Comment by Baterka** [comment]
> While you're at it could you please add an example usage to our `steam-tradeoffer-manager-tests.ts`? Perhaps just some basic class extending from `TradeOffer`.

Oh. Ok. It is basically tested by the code what is already there, but I will try to come up with something.

File: types/steam-tradeoffer-manager/index.d.ts
Comment (kldzj) [inline types/steam-tradeoffer-manager/index.d.ts:14]: Instead of mixing the import styles I'd really prefer having these types live in the namespace.

Comment (Baterka) [reply types/steam-tradeoffer-manager/index.d.ts:14]: Ok

**Comment by kldzj** [comment]
> Oh. Ok. It is basically tested by the code what is already there, but I will try to come up with something.

Really just to verify that the export is working as intended. :)

Thank you very much!

File: types/steam-tradeoffer-manager/lib/classes/TradeOffer.d.ts
Comment (kldzj) [inline types/steam-tradeoffer-manager/lib/classes/TradeOffer.d.ts:6]: FYI: build is failing because of a missing semicolon :)

Comment (Baterka) [reply types/steam-tradeoffer-manager/lib/classes/TradeOffer.d.ts]: Interesting... Not on my machine... :D
Btw why we even need them in JS?  D

Comment (kldzj) [reply types/steam-tradeoffer-manager/lib/classes/TradeOffer.d.ts]: > Btw why we even need them in JS?

We don't necessarily, it's just styling enforcement. :)

Although there can be specific small issues that arise from not using semicolons :P

File: types/steam-tradeoffer-manager/index.d.ts
Comment (kldzj) [inline types/steam-tradeoffer-manager/index.d.ts:655]: You can (and must) omit the `TradeOfferManager` prefix as we're already in this namespace :)

Comment (Baterka) [reply types/steam-tradeoffer-manager/index.d.ts]: Annoying strict rules :D

Comment (kldzj) [reply types/steam-tradeoffer-manager/index.d.ts]: I know the struggle, but imagine the quality of this repo if there were no rules. 😄

**Review by kldzj — APPROVED** [review]
Looks good to me, thank you very much for your contribution 🎉 ❤️

**Comment by Baterka** [comment]
Should I do the test then? Or it will be merged and usable now?

**Comment by kldzj** [comment]
If you were to push a simple test importing and using the `TradeOffer` class in any way you could get rid of this pesky `Untested Change` label 😋

**Comment by kldzj** [comment]
Once you add the test we should be able to merge it ourselves, otherwise a DT maintainer has to look over it.

File: types/steam-tradeoffer-manager/steam-tradeoffer-manager-tests.ts
Comment (kldzj) [inline types/steam-tradeoffer-manager/steam-tradeoffer-manager-tests.ts:158]: You can keep those tests, but it'd be important to add a test for the import using the path so we know that everything is exported as expected.

Comment (Baterka) [reply types/steam-tradeoffer-manager/steam-tradeoffer-manager-tests.ts:158]: Oh.

**Comment by kldzj** [comment]
While writing the test I noticed a possible issue for your use case: the `TradeOffer` constructor is private, meaning it's not extendable.

**Comment by Baterka** [comment]
Yea. You are right... Whole point of this was to make it extendable :D

**Comment by Baterka** [comment]
I hope that commits in merges are squashed :D Bcs this is embarrassing commit history :D

**Comment by kldzj** [comment]
> I hope that commits in merges are squashed :D Bcs this is embarrassing commit history :D

They are. 😋 

> Yea. You are right... Whole point of this was to make it extendable :D

I'm not too sure about simply dropping the private constructor. The TradeOffer class is meant to be created through `manager.createOffer()`. There's a few [properties being set](https://github.com/DoctorMcKay/node-steam-tradeoffer-manager/blob/master/lib/index.js#L397C18-L399).

But in the end the developer would just have to be aware of this.

**Comment by kldzj** [comment]
I think we should be fine dropping it. If anybody is ever trying to instantiate it directly they're not using the documented examples anyway.

**Comment by Baterka** [comment]
Yea... You are right... But I think that everyone will understand it that they need to do `manager.createOffer` and then cast it if they want custom behavior of the TradeOffer, like me. I am planning to just wrap the TradeOffer in something like `TradeOfferPromise` and than make `<method>Async()` methods that promisify the original methods. Is that stupid or make sense? I do not want to use callbacks anymore and also do not want to fork the lib and rewrite it :D

**Comment by kldzj** [comment]
> and then cast it if they want custom behavior of the TradeOffer

I fear this is the problem. It's not enough to overwrite the return type, it would still be the same object.

You'd have to create your own `createCustomOffer` function or method that instantiates the `ExtendedTradeOffer` class and does the same as `manager.createOffer` (set the properties). Another solution could be to "merge" these classes at runtime, but I think that's a bit hackier and much less fun.

**Comment by Baterka** [comment]
And will this do the job at least for now?

**Comment by kldzj** [comment]
> And will this do the job at least for now?

In terms of changes in this PR? Yes, all those other things must be implemented in your repos.

Although, I'd like you to remove line 159 from the tests, as it suggests behavior that will not work. After that I'd be happy to merge this PR. :)

**Comment by kjsmita6** [comment]
Make sure you add this file to the tsconfig.json as well. Then, run `npm test` to make sure everything validates.

**Comment by kldzj** [comment]
> Make sure you add this file to the tsconfig.json as well. Then, run `npm test` to make sure everything validates.

Good catch! But better would probably be to add an glob path for the lib directory (`lib/**/*.d ts`), what do you think?

**Comment by kldzj** [comment]
Although interestingly enough it was tested before, probably because it's imported through `index.d.ts`.

**Comment by kjsmita6** [comment]
Hmm, yeah I see that. I wonder if the tests specifically do that already. Probably best to still add it.

**Comment by Baterka** [comment]
So what I should do?

**Comment by kldzj** [comment]
> So what I should do?

In the `tsconfig.json` add `"lib/**/*.d ts",` into the `files` array, best after `"index.d.ts",`

**Comment by Baterka** [comment]
> > So what I should do?
> 
> In the `tsconfig.json` add `"lib/**/*.d ts",` or the file specific relative path into the `files` array, best after `"index.d.ts",`.
> 
> Oh, and please don't forget to remove line 159 from the tests file, the class extension should be enough as a test :)

2023-05-22T08:54:25.8606788Z Error: steam-tradeoffer-manager: Only index.d.ts may be listed explicitly in tsconfig's "files" entry.

**Comment by kldzj** [comment]
Yeah, it seems we shouldn't do that. Remove it again and it's ready to merge. Sorry for all this hassle, and again many thanks for your contribution. 👍🏻

**Comment by kldzj** [comment]
[Two more minor issues](https://dev.azure.com/definitelytyped/DefinitelyTyped/_build/results?buildId=156880&view=logs&j=8635df7d-ecde-52a7-c7a8-84b998e35690&t=c264d665-0b79-59b7-7e04-9965cbe74860&l=126) and you're good. :)

**Comment by Baterka** [comment]
Should be ok now :D

**Review by kldzj — APPROVED** [review]

**Comment by Baterka** [comment]
Ready to merge

**Pull request merged**
